text/0000-futures.md

```diff
@@ -0,0 +1,617 @@
+
+  - Via a subtrait relationship, so that `T: Future` is defined essentially as
+    an alias for `for<'a> Pin<&mut 'a T>: MoveFuture`. Unfortunately, such
+    "higher ranked" trait relationships don't currently work well in the trait
+    system, and this approach also makes things more convoluted when
+    implementing `Future` by hand, for relatively little gain.
+
+The drawback of the "one core trait" approach taken by this RFC is its ergonomic
+hit when writing moveable futures by hand: you now need to import `Pin` and
+`Unpin`, invoke `poll_unpin`, and impl `Unpin` for your types. This is all
+pretty mechanical, but it's a pain. It's possible that improvements in `Pin`
+ergonomics will obviate some of these issues, but there are a lot of open
+questions there still.
+
+On the other hand, a two-trait approach has downsides as well. If we *also*
+remove the error type, there's a combinatorial explosion, since we end up
+needing `Try` variants of each trait (and this extends to related traits, like
+`Stream`, as well). More broadly, with the one-trait approach, `Unpin` acts as a
+kind of "independent knob" that can be applied orthogonally from other concerns;
+with the two-trait approach, it's "mixed in". And both of the two-trait
+approaches run up against compiler limitations at the moment, though of course
+that shouldn't be taken as a deciding factor.
+
+**The primary reason this RFC opts for the one-trait approach is that it's the
+conservative, forward-compatible option, and has proven itself in practice**.
+It's possible to add `MoveFuture`, together with a blanket impl, at any point in the future.
+Thus, starting with just the single `Future` trait as proposed in this RFC keeps our options
+maximally open while we gain experience.
+
+## Rationale, drawbacks and alternatives to the wakeup design (`Waker`)
+
+Previous iterations of this proposal included a separate wakeup type,
+`LocalWaker`, which was `!Send + !Sync` and could be used to implement
+optimized executor behavior without requiring atomic reference counting
+or atomic wakeups. However, in practice, these same optimizations are
+available through the use of thread-local wakeup queues, carrying IDs
+rather than pointers to wakeup objects, and tracking an executor ID or
+thread ID to perform a runtime assertion that a `Waker` wasn't sent across
+threads. For a simple example, a single thread-locked executor with zero
+atomics can be implemented as follows:
+
+```rust
+struct Executor {
+    // map from task id (usize) to task
+    tasks: Slab<Task>,
+    // list of woken tasks to poll
+    work_queue: VecDeque<usize>,
+}
+
+thread_local! {
+  pub static EXECUTOR: RefCell<Option<Executor>> = ...;
+}
+
+static VTABLE: &RawWakerVTable = &RawWakerVTable {
+    clone: |data: *const ()| RawWaker { data, vtable: VTABLE, },
+    wake: |data: *const ()| EXECUTOR.borrow_mut().as_mut().expect(...).work_queue.push(data as usize),
+    drop: |_: *const ()| {},
```

Comment:
I recognized the inconsistencies before, when @aturon initially proposed those names. And afaik I tried to remove the `_fn` from `drop_fn`. However I run into some compilation errors (reserved name or something like this), so I had to add it again.
Haven't tried it again recently. Maybe things changed with the latest Rust 2018 releases?

Reply:
Drop is not a reserved name; otherwise it would not be possible to define `fn drop<T>(_: T) {}` in the prelude. Do you remember where the errors arose from? The struct definition? Somewhere else?

Reply:
I unfortunately can't remember what exactly was wrong. Maybe I can port my old changes to the new API on the weekend and try to reproduce it.

Reply:
Cool, thanks ❤️

Reply:
Seems like it works with pure `drop` too, so we can rename it.
I'm personally fine with both.

Reply:
We should be consistent here, either all these fn pointer fields end with `_fn` or none do (I prefer none).